rust/src/encodings/plain.rs

```diff
@@ -39,50 +43,49 @@ impl<'a, R: Read + Seek> PlainDecoder<'a, R> {
     }
 }
 
-impl<'a, R: Read + Seek, T: ArrowPrimitiveType> Decoder<T> for PlainDecoder<'a, R> {
+impl<'a, R: Read + Seek, T: NativeType> Decoder<T> for PlainDecoder<'a, R> {
     type ArrowType = T;
 
-    fn decode(&mut self, offset: i32, length: &Option<i32>) -> Result<ArrayRef> {
+    fn decode(&mut self, offset: i32, length: &Option<i32>) -> Result<Arc<dyn Array>> {
         let read_len = length.unwrap_or((self.page_length - (offset as i64)) as i32) as usize;
         (*self.file).seek(SeekFrom::Start(self.position + offset as u64))?;
-        let mut mutable_buf = MutableBuffer::new(read_len * T::get_byte_width());
-        (*self.file).read_exact(mutable_buf.as_slice_mut())?;
-        let builder = ArrayDataBuilder::new(T::DATA_TYPE).buffers(vec![mutable_buf.into()]);
-        let data = match builder.build() {
-            Ok(d) => d,
-            Err(e) => {
-                return Err(Error::new(
-                    ErrorKind::InvalidData,
-                    format!("Invalid builder: {}", e),
-                ))
+        // let mut mutable_buf = Buffer::new(read_len * T::get_byte_width());
+        let byte_size = std::mem::size_of::<T>();
+        let mut buf = vec![0u8; read_len * byte_size];//TODO is it right?
+        (*self.file).read_exact(&mut buf)?;
+        let mut builder = MutablePrimitiveArray::with_capacity(read_len);
```

Comment:
is there an arrow2 api to directly pass the continuous memory buffer into Array?

This assignment of each element is not efficient.

Reply:
I don't see the API that can batch convert raw bytes to PrimitiveArray in Arrow2. If we want batch convert, maybe we can use [ReadBytesExt](https://docs.rs/byteorder/latest/byteorder/trait.ReadBytesExt.html)? The code would become similar to blow

```
        match T::PRIMITIVE {
            PrimitiveType::Int8 => {
                let mut vec = vec![0i8; read_len];
                (*self.file).read_i8_into(&mut vec)?;
                PrimitiveArray::from_slice(vec).arced();
            }
            PrimitiveType::Int16 => {
                let mut vec = vec![0i16; read_len];
                (*self.file).read_i16::<LittleEndian>(&mut vec)?;
                PrimitiveArray::from_slice(vec).arced();
            }
            /**
            Boilerplate code ...
            **/
            PrimitiveType::Int32 => {}
            PrimitiveType::Int64 => {}
            PrimitiveType::Int128 => {}
            PrimitiveType::Int256 => {}
            PrimitiveType::UInt8 => {}
            PrimitiveType::UInt16 => {}
            PrimitiveType::UInt32 => {}
            PrimitiveType::UInt64 => {}
            PrimitiveType::Float16 => {}
            PrimitiveType::Float32 => {}
            PrimitiveType::Float64 => {}
            PrimitiveType::DaysMs => {}
            PrimitiveType::MonthDayNano => {}
        }
```

Reply:
This way is not applicable for `DaysMs` and `MonthDayNano`.

Reply:
The reason that we are using arrow is that arrow is a "zero-copy" representation of arrays. Copying / constructing an array should be just setting a few pointers. 

https://arrow.apache.org/docs/format/Columnar.html

Reply:
So as an implementation of Arrow, Arrow2 should definitely have a function that can convert `[u8]` to `PrimitiveArray` without copy data? I would take a further look into it.

Reply:
One more question, Arrow2 said it's [transmute-free](https://github.com/jorgecarleitao/arrow2).
In my understanding, the meaning of transmute-free is:
We don't provide such a way that "Copying / constructing an array should be just setting a few pointers."
Wrong?

Reply:
I've found some very similar question in the Arrow2 [issues](https://github.com/jorgecarleitao/arrow2/issues/987#issuecomment-1123229136), will copy from it.